### PR TITLE
Local transpose in Analyzer

### PR DESCRIPTION
**Pull request by xinrong-meng**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -1470,6 +1470,22 @@ case class Pivot(
...
 
+case class Transpose(
+    indexColumn: Expression,
+    child: LogicalPlan,
+    originalColNames: Option[Seq[String]] = None,
+    output: Seq[Attribute] = Seq.empty
+) extends UnaryNode {
+
+  override lazy val resolved: Boolean = {
+    indexColumn.resolved && child.resolved && originalColNames != null && output.nonEmpty
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]:
```suggestion
    indexColumn.resolved && child.resolved && originalColNames.isDefined && output.nonEmpty
```

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2173,6 +2173,37 @@ class Dataset[T] private[sql](
...
+   *
+   * @param indexColumn The single column that will be treated as the index for the transpose
+   * operation.This column will be used to pivot the data, transforming the DataFrame such that
+   * the values of the indexColumn become the new columns in the transposed DataFrame. If not
+   * provided, the first column is used.
+   * 
+   * @group untypedrel
+   * @since 4.0.0
+   */
+  def transpose(indexColumn: Option[Column] = None): DataFrame = withPlan {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: It's better to have two overloads, one with indexColumn parameter and one without.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      }
+
+      // Cast the index column to StringType
+      val indexColumnAsString = Alias(
+        Cast(indexColumn, StringType),
+        indexColumn.asInstanceOf[Attribute].name)().asInstanceOf[NamedExpression]
+
+      // Cast non-index columns to the least common type
+      val nonIndexColumnsAttr = child.output.filterNot(
+        _.exprId == indexColumn.asInstanceOf[Attribute].exprId)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: what if index column is not an attribute? such as a nested field like `a.b`

**Comment by cloud-fan** [comment]
I'm a bit worried about observability here as when users look at the analyzed plan, they only see `LocalRelation`. My suggestion is to start with `UnresolvedTranspose`, and then resolve it to `Transpose` with validation applied, cast added, and output attributes specified. In the beginning of optimizer, the rule of `FinishAnalysis`, we execute the transpose and convert it to LocalRelation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 1 above)
Comment (xinrong-meng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Makes sense, thank you!

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Used overloads instead, thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 3 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: I modified to enforce index column to be an attribute (of AtomicType)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 4]
```diff
@@ -925,3 +925,15 @@ case object UnresolvedWithinGroup extends LeafExpression with Unevaluable {
   override def dataType: DataType = throw new UnresolvedException("dataType")
   override lazy val resolved = false
 }
+
+case class UnresolvedTranspose(
+    indexColumn: Expression,
+    child: LogicalPlan,
+    originalColNames: Option[Seq[String]] = None,
+    override val output: Seq[Attribute] = Seq.empty
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: unresolved node does not need to define `output`, see also `UnresolvedRelation`

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        finalMatrix(i)(j) = transposedMatrix(i)(j - 1)
+      }
+    }
+
+    finalMatrix
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(TreePattern.UNRESOLVED_TRANSPOSE)) {
+    case t @ UnresolvedTranspose(indexColumn, child, _, _) if !t.resolved =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]:
```suggestion
    case t @ UnresolvedTranspose(indexColumn, child, _, _) if indexColumn.resolved && child.resolved =>
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 6]
```diff
@@ -925,3 +925,15 @@ case object UnresolvedWithinGroup extends LeafExpression with Unevaluable {
   override def dataType: DataType = throw new UnresolvedException("dataType")
   override lazy val resolved = false
 }
+
+case class UnresolvedTranspose(
+    indexColumn: Expression,
+    child: LogicalPlan,
+    originalColNames: Option[Seq[String]] = None,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala]: This field is not needed?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        val keyAttr = AttributeReference("key", StringType, nullable = false)()
+        val transposedColumnNames = fullCollectedRows.map { row => row.getString(0) }
+        val valueAttrs = transposedColumnNames.map { value =>
+          AttributeReference(
+            value,
+            commonType
+          )()
+        }
+
+        LocalRelation(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: Again, we should not turn it into `LocalRelation` during analysis, as we need observability. Let's split your code here into two parts: add casts and collect results to do transpose.

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 5 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: Adjusted.

May I ask why not `!t.resolve`?

File: sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala (hunk 7 above)
Comment (xinrong-meng) [reply sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala]: Sounds good, done.

**Comment by xinrong-meng** [comment]
Please see https://github.com/xinrong-meng/spark/pull/5

**Pull request closed without merging**